pkg/tools/etcd_helper.go

```diff
@@ -116,19 +120,51 @@ func (h *EtcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 			}
 			continue
 		}
-		obj := reflect.New(v.Type().Elem())
-		if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
-			return err
-		}
-		if h.Versioner != nil {
-			// being unable to set the version does not prevent the object from being extracted
-			_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+		if obj, found := h.GetFromCache(node.ModifiedIndex); found {
+			v.Set(reflect.Append(v, obj.Elem()))
+		} else {
+			obj := reflect.New(v.Type().Elem())
+			if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
+				return err
+			}
+			if h.Versioner != nil {
+				// being unable to set the version does not prevent the object from being extracted
+				_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+			}
+			v.Set(reflect.Append(v, obj.Elem()))
+			if node.ModifiedIndex != 0 {
+				h.AddToCache(node.ModifiedIndex, obj)
+			}
 		}
-		v.Set(reflect.Append(v, obj.Elem()))
 	}
 	return nil
 }
 
+type EtcdCache interface {
+	GetFromCache(index uint64) (obj reflect.Value, ok bool)
+	AddToCache(index uint64, obj reflect.Value)
+}
+
+func (h *EtcdHelper) GetFromCache(index uint64) (obj reflect.Value, ok bool) {
+	h.mutex.RLock()
+	defer h.mutex.RUnlock()
+	obj, ok = h.cache[index]
```

Comment:
Would need to do a copy here to prevent people from mutating the cache - add that so your benchmarks are accurate.